ExtraDry/ExtraDry.Server/RuleEngine.cs

```diff
@@ -185,42 +146,195 @@ public UndeleteResult Undelete<T>(T item)
+    /// <param name="remove">The action that is executed for a hard-delete.</param>
+    /// <param name="commit">The action that is executed to commit hard-delete.</param>
+    [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Keep as standard service instance style for DI.")]
+    public async Task<DeleteResult> TryHardDeleteAsync<T>(T item, Func<Task> remove, Func<Task> commit)
+    {
+        ArgumentNullException.ThrowIfNull(item, nameof(item));
+
         if(remove == null) {
             throw new ArgumentNullException(nameof(item), "Must provide an action to prepare hard delete, e.g. remove from collection");
         }
+
         if(commit == null) {
             throw new ArgumentNullException(nameof(item), "Must provide an action to commit hard delete, e.g. save to database");
         }
         var result = DeleteResult.NotDeleted;
-        if(AttemptSoftDelete(item)) {
-            await commit();
-            result = DeleteResult.SoftDeleted;
-        }
+
         try {
             await remove();
             await commit();
             result = DeleteResult.HardDeleted;
         }
-        catch { // Hmmm, can we be specific about exception type here?
-            if(result == DeleteResult.NotDeleted) {
-                throw new InvalidOperationException("Item could not be hard-deleted and no rules for soft-delete provided.");
+        catch {
+            //Do not throw exceptions on commit- Just return appropriate result.
+        }
+        return result;
+    }
+
+    /// <summary>
+    /// Register the remove action to be used when hard-deleting an item.
+    /// </summary>
+    /// <param name="remove">The action that is executed for a hard-delete.</param>
+    public void RegisterRemove<T>(Action<T> remove)
+    {
+        RemoveFunctors.Add(typeof(T), e => remove((T)e));
+    }
+
+    /// <summary>
+    /// Register the commit action to be used when hard-deleting an item.
+    /// </summary>
+    /// <param name="commit">The action that is executed to commit hard-delete.</param>
+    public void RegisterCommit(Func<Task> commit)
+    {
+        CommitFunctor = commit;
+    }
+
+    /// <inheritdoc cref="TryHardDeleteAsync(object[])" />
+    public async Task<DeleteResult> TryHardDeleteAsync<T>(T item) where T : class
+    {
+        Action<object> remove;
+        try {
+            remove = RemoveFunctors[typeof(T)];
+        }
+        catch(KeyNotFoundException ex) {
+            throw new DryException($"Could not find registered expression for removing Entity. Register Entity removal with the RegisterRemove method.", ex);
+        }
+
+        var result = DeleteResult.NotDeleted;
+
+        try {
+            remove(item);
+            await CommitFunctor();
+            result = DeleteResult.HardDeleted;
+        }
+        catch {
+            //Do not throw exceptions on commit - Just return appropriate result.
+        }
+        RemoveFunctors.Clear();
```

Comment:
What is the reason to clear these here?  I can't think of a reason why they are needed, but if I was going to call another hard delete I don't think I would expect my registrations to have gone missing.

Reply:
This was just a cleanup but registrations going missing on another call is a good point - removed.